### PR TITLE
build(docs-infra): improve search quality

### LINKED ISSUE #24380: provide more cues for deprecated APIs
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[ ] Feature request
[X ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ X] Other... Please describe:  info arch (content access) and AIO infra
</code></pre>

## Current behavior

1. Search results: If I search for "http", the search results include (and sometimes favor) the deprecated APIs. This is good, because search should provide complete results. This is not so good, because it can result in 4 steps instead of 2 steps to find the current API:

- Search for http
- Pick "http" from search results
- Discover that it's deprecated, so it's probably not what I want
- Click link to replacement API

deprecated-http-search

2. API landing page: If I am on the API landing page, all APIs are shown without a status indicator. I can filter for only stable, only experimental, or only deprecated, but without filtering I can't tell which ones are deprecated. Sometimes the deprecated API is higher in the index or a better match for my keyword, which means I'm likely to have extra clicks:

- Go to API index
- Type "http" into search field. 
- Choose "http"
- Find out that it's' deprecated. Follow link to replacement API

deprecated-http-api-index

3. Module page: No indicator if module is deprecated
https://angular.io/api/http
deprecated-http-module

## Expected behavior

1. It might be useful to append status to APIs in the search results for APIs that are not stable, like this: 

http (deprecated)
animations (experimental)
http client

Or shorthand that users can learn easily

http (dep)(old?)
animations (exp)
http client

And/or filter search results automatically to put stable APIs at the top of the results. 

2. Similar to #1, append status to API names in the list, and/or offer a filter that excludes deprecated (so I can see stable + experimental together, which is everything I should consider for possible future direction). 

3. Add "deprecated" badge to deprecated packages, as mocked up in image above. 

## Minimal reproduction of the problem with instructions

See above. 

## What is the motivation / use case for changing the behavior?

Reduce navigation "noise" and extra steps for developers. Make non-deprecated APIs more prominent in navigation paths. 

## Environment

<pre><code>
Angular version: X.Y.Z

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX  
- Platform:  

Others:

</code></pre>

### LINKED ISSUE #25721: Indicate deprecated APIs in package overview pages
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[ ] Feature request
[ X] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[X ] Other... Please describe: API doc templates 
</code></pre>

## Current behavior

In the package overview pages, there is no indication of which exports are deprecated. This seems like a very handy place to let developers know what to avoid using. 

## Expected behavior

Case 1: Entire package is deprecated. Should have [deprecated] icon at top and a clear note somewhere near the top, with info and link to replacement. 

Case 2: Some exports are deprecated. To be designed. One option is simply begin the description with "Deprecated."

## Minimal reproduction of the problem with instructions

Example: https://angular.io/api/http
Almost all of those exports are deprecated.

## What is the motivation / use case for changing the behavior?

Make it easier for developers to make good choices. Avoid having to click into each export's page to see that it's deprecated. 

## Environment

<pre><code>
Angular version: X.Y.Z

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX  
- Platform:  

Others:

</code></pre>

### PR DESCRIPTION
**Pull request by petebacondarwin**
Fixes #24380
Fixes #25721

### CODE REVIEW

**Comment by IgorMinar** [comment]
This seems to be better. @jenniferfell can you test your queries with this version please?

**Review by gkalpak — APPROVED** [review]

File: aio/tools/transforms/angular-base-package/processors/generateKeywords.js [hunk 1]
```diff
@@ -52,20 +49,33 @@ module.exports = function generateKeywordsProcessor(log, readFilesProcessor) {
...
-        return text.replace(/(^|\s)([nN]g([A-Z]\w*))/g, '$1$2 $3');
+      function tokenize(text) {
+        const rawTokens = text.split(/[\s\/]+/mg);
+        const tokens = [];
+        rawTokens.forEach(token => {
+          // Strip off unwanted trivial characters
+          token = token
+              .trim()
+              .replace(/^[_\-"'`({[<$*]+/, '')
+              .replace(/[_\-"'`)}\]>.]+$/, '');
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply aio/tools/transforms/angular-base-package/processors/generateKeywords.js]: Hmm, good spot. This means that we might end up with things like `foo(` in the index...

[the remark this answers is not in the document]
Comment (petebacondarwin) [reply aio/tools/transforms/angular-base-package/processors/generateKeywords.js]: In the query lexers (on the browser) we now split the terms by whitespace only. So the following query, https://pr25750-f2fa79e.ngbuilds.io/?search=downgradeModule(), results in no terms matching. Given that we will not keep such punctuation in the index, we should modify the lexer appropriately.

**Comment by petebacondarwin** [comment]
We considered just zipping up and downloading the entire text content of the docs, then building the index directly from the original content. This is problematic in a number of ways:

* the rendered content hides structural information such as headings, titles and most importantly members of classes
* the parsing of the rendered content does not result in improvements to the searching issues highlighted
* the size of the content (even zipped) is bigger than the current download
* the building of the index is considerably longer (from ~3secs to up to 10secs!).

File: aio/src/app/search/search-worker.js [hunk 2]
```diff
@@ -86,10 +87,13 @@ function loadIndex(searchInfo /*: SearchInfo */) {
...
-      // E.g. if the search is "ngCont guide" then we search for "ngCont guide titleWords:ngCont*"
-      var titleQuery = 'titleWords:*' + query.split(' ', 1)[0] + '*';
-      var results = index.search(query + ' ' + titleQuery);
+      var results = index.search(query);
+      if (results.length === 0) {
+        // Add a relaxed search in the title for the first word in the query
+        // E.g. if the search is "ngCont guide" then we search for "ngCont guide titleWords:ngCont*"
+        var titleQuery = 'titleWords:*' + query.split(' ', 1)[0] + '*';
+        results = index.search(query + ' ' + titleQuery);
+      }
```
Comment (petebacondarwin) [inline aio/src/app/search/search-worker.js]: Basically we now attempt a straight search with the given query; and only fall back on a more relaxed title based search if the original query returns nothing. This results in fewer false positives.

File: aio/src/app/search/search-worker.js [hunk 3]
```diff
@@ -27,12 +27,13 @@ self.onmessage = handleMessage;
 // Create the lunr index - the docs should be an array of objects, each object containing
 // the path and search terms for a page
 function createIndex(addFn) {
+  lunr.QueryLexer.termSeparator = lunr.tokenizer.separator = /\s+/;
```
Comment (petebacondarwin) [inline aio/src/app/search/search-worker.js]: The built-in lexer was splitting on some punctuation, which are actually significant when searching code. This was causing false negatives.  Now we just split on whitespace.

File: aio/tools/transforms/angular.io-package/index.js [hunk 4]
```diff
@@ -37,7 +37,7 @@ module.exports = new Package('angular.io', [gitPackage, apiPackage, contentPacka
     checkAnchorLinksProcessor.$runBefore = ['convertToJsonProcessor'];
     checkAnchorLinksProcessor.$runAfter = ['fixInternalDocumentLinks'];
     // We only want to check docs that are going to be output as JSON docs.
-    checkAnchorLinksProcessor.checkDoc = (doc) => doc.path && doc.outputPath && extname(doc.outputPath) === '.json';
+    checkAnchorLinksProcessor.checkDoc = (doc) => doc.path && doc.outputPath && extname(doc.outputPath) === '.json' && doc.docType !== 'json-doc';
```
Comment (petebacondarwin) [inline aio/tools/transforms/angular.io-package/index.js]: If we don't ignore json-doc files, the anchor checker get confused by the JSON escaped URLs in the rendered content of the search term file.

**Comment by petebacondarwin** [comment]
@jenniferfell has suggested that supporting quoted "phrase searching", to get fewer false positives, but lunr does not support this out of the box, and effort required to implement it on top of lunr, the benefit seemed unworthy.

**Comment by petebacondarwin** [comment]
Still to be done (in another PR):

* lower the ranking of deprecated API docs

File: aio/tools/transforms/angular-base-package/processors/generateKeywords.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply aio/tools/transforms/angular-base-package/processors/generateKeywords.js]: Fixed

**Comment by petebacondarwin** [comment]
Here are some screenshots of the changes...

screen shot 2018-09-17 at 19 28 11
screen shot 2018-09-17 at 19 27 32
screen shot 2018-09-17 at 19 27 12
screen shot 2018-09-17 at 19 26 56

**Comment by petebacondarwin** [comment]
Personally I think the dark grey works better for the deprecated badge, since it suggests something is outdated and old rather than a serious issue.

**Comment by petebacondarwin** [comment]
Note that entry points that have all their exports deprecated are now also deprecated.

**Comment by IgorMinar** [comment]
- [ ] remove "(deprecated)" from search
- [ ] let's ensure that packages like @angular/upgrade are not marked as deprecated (package page doesn't have a badge unless all entry points are deprecated)
- [ ] add strike through to the api listing

**Review by IgorMinar — APPROVED** [review]

**Comment by jenniferfell** [comment]
**Site search for http:** 
- API results are much better. Putting all the deprecated results last is a big improvement.
- API results with strikethrough for deprecated APIs LGTM. (fwiw...for others reading this who might not know...Google Dev Doc Style Guide doesn't include this convention, nor do Android dev docs, but I do see it in xcode and other 3rd party docs. It's an emerging convention, and it's much cleaner than the word "deprecated" in the results list.)

- _Hmmm....Guide results are odd._ HTTPClient is 3rd in the results. But at least still in the top 3. Not sure if anything can be done on this without breaking something else. We might need to look at why and then maybe tweak content of guides slightly? Is this because matches in titles aren't taking precedence any more?

**Site search for node.js**
- Celebration! :-) Perfect!

**API index page:**
- Strikethrough of deprecated LGTM
- When package type selected, disabled status field LGTM

**Badges:**
- Having "deprecated" back again is great, but it would be nice to use a more alert-like color than gray. Can we switch and have ngmodules be gray and deprecated be the red that ngmodules is right now? Or maybe an orange, because security risk is (beautifully and perfectly) red?

**API package and type pages for deprecated APIs:**
- LGTM. Pages for types such as HttpModule page LGTM. I like the deprecated badge and the deprecated bold in-line title at the end of the description.
- Package overview pages. Having strikethrough and Deprecated in bold for things that are deprecated is great. It's a bit heavy when the entire package is deprecated, but I prefer keeping that for consistency...and I really like having it in packages like Common where only a few things are deprecated. 


No show stoppers. MUCH IMPROVED!!

**Review by jenniferfell — APPROVED** [review]

**Comment by petebacondarwin** [comment]
Great. I will fix the broken Travis test and swap the colours. Then mark it for merge.

**Comment by petebacondarwin** [comment]
Regarding colour changes. This is what I came up with:

screen shot 2018-09-20 at 19 47 13

**Comment by JoostK** [comment]
There's something weird going on with the search results, it seems.

I was exploring the various modules to see their badge color and started searching for `ReactiveFormsModule`. If you arrive at typing `React` (I'm very sorry 😆) then all of a sudden `NgElementStrategy` is the only result, whereas one character less, or more, will resurface `ReactiveFormsModule`. The current deployment on angular.io does not show the same behaviour.

Regardless of this minor thing, this PR brings a huge improvement to the search results overall!

**Comment by petebacondarwin** [comment]
@JoostK thanks for testing this. I can explain *why* this is happening but I am not sure we can do much better. Here is what happens...

* `reac`
  - The stemmer does not convert `reac` into any matching keyword
  - So no results from a normal query
  - trigger the special title query, which is much more relaxed and matches any chars in the title
  - `ReactiveFormsModule` API page and `Reactive Forms` guide page match
* `react`
  - The stemmer does convert this to match keywords such as `react` and `reacts`, but not `reactive` it seems.
  - `NgElementStrategy` contains the word `react` in its text, so is included in the results
  - `Component Interaction` contains the word `reacts` in its text, so is included in the results
  - `ReactiveFormsModule` only contains the word `reactive` and not `react` or `reacts` etc, so is included in the results
* `reacti`
  - once again the stemmer fails to match keywords and so we are back to the special title query again, with the same results as for `reac`.

We don't have any control over the stemmer, AFAIK, so there is not so much we can do.

**Pull request closed without merging**